### PR TITLE
new feature: remove a specific song from somewhere in the queue

### PR DESCRIPTION
**Pull request by ghost**
I tried to match the feel of the other commands. I also made some pull requests that I had to close because I've never used GitHub before half an hour ago. Oops!

This removes a song from somewhere in the queue, tells you what song was removed, and doesn't take bad input.

### CODE REVIEW

**Comment by CArchetype** [comment]
Need to change () to {} for the !help comment to keep consistent with other commands. Will do that next time I'm on PC.

File: musicbot/bot.py [hunk 1]
```diff
@@ -1702,6 +1702,47 @@ def check(m):
...
+            {command_prefix}remove [number]
+	    
+        Removes a song from the queue at the given position, where the position is a number from {command_prefix}queue.
+        """
+
+        # if the queue has as many entries as the number they
+        # input, we know there's a song there
+        # if the user inputs 0, tell them to use skip instead
+        if not index:
+            raise exceptions.CommandError("No input found. Try again with a position from {command_prefix}queue.", expire_in=20)
```
Comment (jayktaylor) [inline musicbot/bot.py]: You need to format `{command_prefix}` yourself here, the exception will not format it for you.

File: musicbot/bot.py [hunk 2]
```diff
@@ -1702,6 +1702,47 @@ def check(m):
                 delete_after=20
             )
 
+    async def cmd_remove(self, message, player, index=None):
```
Comment (jayktaylor) [inline musicbot/bot.py]: `index` should be given as a positional argument, not a kwarg. It is a required argument for the command to work, right?

File: musicbot/bot.py [hunk 3]
```diff
@@ -1702,6 +1702,47 @@ def check(m):
...
+                except IndexError:
+                    raise exceptions.CommandError("Something went wrong. Please try again.", expire_in=20)
+                
+                return Response(':white_check_mark: **' + song_title + '** removed.', delete_after=20)
+
+            else:
+                raise exceptions.CommandError("You can't remove the current song (skip it instead), or a song in a position that doesn't exist.", expire_in=20)
+
+        else:
+            raise exceptions.CommandError("There is either no queue, or the queue is not that long.", expire_in=20)
```
Comment (jayktaylor) [inline musicbot/bot.py]: Maybe it's worth checking if there are entries in the queue separately above this entire `if/else`.

File: musicbot/bot.py [hunk 4]
```diff
@@ -1702,6 +1702,47 @@ def check(m):
...
+            raise exceptions.CommandError("No input found. Try again with a position from {command_prefix}queue.", expire_in=20)
+
+        try:
+            index = int(index)
+
+        except ValueError:
+            raise exceptions.CommandError('{} is not a valid number.'.format(index), expire_in=20)
+
+        if index <= len(player.playlist.entries):
+            if index > 0:
```
Comment (jayktaylor) [inline musicbot/bot.py]: Why not group these two `if` statements onto the same line?

File: musicbot/bot.py [hunk 5]
```diff
@@ -1702,6 +1702,47 @@ def check(m):
...
+                try:
+                    player.playlist.remove_entry((index)-1)
+
+                # this catches the rare case where the player
+                # loads a song and we try to remove a position on
+                # the very end of the queue that no longer exists
+                except IndexError:
+                    raise exceptions.CommandError("Something went wrong. Please try again.", expire_in=20)
+                
+                return Response(':white_check_mark: **' + song_title + '** removed.', delete_after=20)
```
Comment (jayktaylor) [inline musicbot/bot.py]: Copy the way we embed emojis like that in strings: https://github.com/SexualRhinoceros/MusicBot/blob/review/musicbot/bot.py#L1035

**Comment by jayktaylor** [comment]
Accidental assignment, woops! Looks better 👍 Bork will probably take a look in future after he's got all of the voice stuff out of the way.

**Comment by zstinnett** [comment]
I added this to my Musicbot and I love it!

**Comment by KingGold171** [comment]
code as is.

**Comment by Ylsovr** [comment]
How do I download this? Im new sorry :/

**Comment by pedrofracassi** [comment]
Seems to be working fine here, nice feature!

**Pull request merged**